[Fix] Unable to build on install by use the right lifecircle hook

A0nameless0man opened the pull request.

Nugine commented:
根据 <https://docs.npmjs.com/cli/v7/using-npm/scripts>，prepack 是正解。
yarn v1 有 bug，不会运行 prepack，但这个 bug 不会被修复 <https://github.com/yarnpkg/yarn/issues/5235#issuecomment-754993583>
用 npm 配合 prepack 可以正常自动编译。

原来的 format 和 lint 影响的文件范围有错，这个应该修一下。

A0nameless0man commented:
> 根据 https://docs.npmjs.com/cli/v7/using-npm/scripts，prepack 是正解。
> yarn v1 有 bug，不会运行 prepack，但这个 bug 不会被修复 [yarnpkg/yarn#5235 (comment)](https://github.com/yarnpkg/yarn/issues/5235#issuecomment-754993583)
> 用 npm 配合 prepack 可以正常自动编译。
> 
> 原来的 format 和 lint 影响的文件范围有错，这个应该修一下。

测试表明，npm安装依赖会在grpc这个包上报错。

Nugine commented:
> > 根据 [https://docs.npmjs.com/cli/v7/using-npm/scripts，prepack](https://docs.npmjs.com/cli/v7/using-npm/scripts%EF%BC%8Cprepack) 是正解。
> > yarn v1 有 bug，不会运行 prepack，但这个 bug 不会被修复 [yarnpkg/yarn#5235 (comment)](https://github.com/yarnpkg/yarn/issues/5235#issuecomment-754993583)
> > 用 npm 配合 prepack 可以正常自动编译。
> > 原来的 format 和 lint 影响的文件范围有错，这个应该修一下。
> 
> 测试表明，npm安装依赖会在grpc这个包上报错。

为啥，协议库没用到grpc吧。来点复现手段。

A0nameless0man commented:
稍等，我开个vultr实例，在干净的环境中测试

A0nameless0man commented:
emmm，测试结果：没报错，但是也没编译
```
[root@vultrguest heng-controller]# rm package-lock.json ./node_modules/ -rf
[root@vultrguest heng-controller]# cat ./package.json | grep heng
    "name": "heng-controller",
        "heng-protocol": "git+https://github.com/ThinkSpiritLab/Heng-Protocol.git",
[root@vultrguest heng-controller]# npm install
//skip a lot
added 1056 packages from 709 contributors and audited 1060 packages in 45.327s

86 packages are looking for funding
  run `npm fund` for details

found 0 vulnerabilities

[root@vultrguest heng-controller]# ls ./node_modules/heng-protocol/
docs  external-protocol  index.ts  internal-protocol  LICENSE  package.json  README.md  tsconfig.json
[root@vultrguest heng-controller]# ls ./node_modules/heng-protocol/internal-protocol/
http.ts  index.ts  ws.ts
[root@vultrguest heng-controller]# git status
On branch dev-use-npm
Your branch is up to date with 'origin/dev-use-npm'.
```

A0nameless0man commented:
使用 `install` 则可以正常编译
```
[root@vultrguest heng-controller]# npm install -save https://github.com/ThinkSpiritLab/Heng-Protocol.git#7ed9fda

> heng-protocol@0.0.4 install /root/heng-controller/node_modules/heng-protocol
> tsc

npm WARN optional SKIPPING OPTIONAL DEPENDENCY: fsevents@2.3.2 (node_modules/fsevents):
npm WARN notsup SKIPPING OPTIONAL DEPENDENCY: Unsupported platform for fsevents@2.3.2: wanted {"os":"darwin","arch":"any"} (current: {"os":"linux","arch":"x64"})

+ heng-protocol@0.0.4
added 1 package from 1 contributor, updated 1 package and audited 1061 packages in 33.031s

86 packages are looking for funding
  run `npm fund` for details

found 0 vulnerabilities

[root@vultrguest heng-controller]# ls ./node_modules/heng-protocol/
docs  external-protocol  index.d.ts  index.js  index.ts  internal-protocol  LICENSE  node_modules  package.json  README.md  tsconfig.json
```

A0nameless0man commented:
又及，我试了一下，确实没有grpc

A0nameless0man commented:
我看了一下文档，觉得应该是 `prepare` 或 `install`

https://docs.npmjs.com/cli/v7/using-npm/scripts#npm-install

Nugine commented:
我们最终是要发布的，发布前手动编译，目前用 git 依赖只是临时手段，所以针对 git 依赖加自动编译是合适的。

之前改的时候忘记我有个全局 tsc 了，这锅我背。。。

这是修复方法 <https://github.com/Nugine/Heng-Protocol/commit/9328770d1e551de3a73ea43a961636d74074dc3d>
把依赖改成 <https://github.com/Nugine/Heng-Protocol> 可以看效果

Nugine commented:
prepare 和 install 都会在协议仓库 `npm install` 时执行编译，产生不必要的垃圾文件。

A0nameless0man commented:
> 我们最终是要发布的，发布前手动编译，目前用 git 依赖只是临时手段，所以针对 git 依赖加自动编译是合适的。
> 
> 之前改的时候忘记我有个全局 tsc 了，这锅我背。。。
> 
> 这是修复方法 [Nugine@9328770](https://github.com/Nugine/Heng-Protocol/commit/9328770d1e551de3a73ea43a961636d74074dc3d)
> 把依赖改成 https://github.com/Nugine/Heng-Protocol 可以看效果

无效果

Nugine commented:
> > 我们最终是要发布的，发布前手动编译，目前用 git 依赖只是临时手段，所以针对 git 依赖加自动编译是合适的。
> > 之前改的时候忘记我有个全局 tsc 了，这锅我背。。。
> > 这是修复方法 [Nugine@9328770](https://github.com/Nugine/Heng-Protocol/commit/9328770d1e551de3a73ea43a961636d74074dc3d)
> > 把依赖改成 https://github.com/Nugine/Heng-Protocol 可以看效果
> 
> 无效果

用 root 安装会有问题，换成普通用户 <https://stackoverflow.com/questions/18136746/npm-install-failed-with-cannot-run-in-wd>

A0nameless0man commented:
可以了
我改动我这边吧

A0nameless0man commented:
已经进行了相应的改动

Nugine commented on package.json:
```diff
@@ -7,19 +7,21 @@
+        "prepack": "npm install typescript && npx tsc"
+    },
+    "dependencies": {
+        "typescript": "^4.1.3"
```
这里不应该依赖 ts，因为未来以 js 形式发布时不需要 tsc

A0nameless0man replied:
修了

A0nameless0man commented:
草，这个commit读起来像是存在只有Nugine安装时会出问题

Nugine commented:
> 草，这个commit读起来像是存在只有Nugine安装时会出问题

emmm
